### DISCUSSION TITLE
Is there a simple script to separate more than one image on the same image file into 2 or more separate images?

### DISCUSSION THREAD

**Discussion by dmcnaugh15** [ImageMagick/ImageMagick#7852] [Help] [answered]
### ImageMagick version

7.1.1-41-Q16-HDRI-x64-dll.exe

### Operating system, version and so on

Windows 11

### Description

I have several scanned image files that contain more than one photo on the image (see example).  Is there any way to separate them into separate image files?

EXAMPLE:
![Example](https://github.com/user-attachments/assets/c6f2ecf1-3e30-4448-bd85-a054fe40edf6)

**Comment by fmw42** [comment]
You can use -connected-components processing after thresholding on your white border to separate the images. See https://imagemagick.org/script/connected-components.php

```
Theshold the image
Connected Components on the thresholded image to get the crop boundaries
For loop to crop the original
```

Unix Syntax

```
magick image.jpg -color-threshold "srgb(240,240,240)-srgb(255,255,255)" -negate -bordercolor black -border 1 x.png

bboxArr=(`magick x.png \
-type bilevel \
-define connected-components:verbose=true \
-define connected-components:keep-top=3 \
-define connected-components:exclude-header=true \
-connected-components 4 null: | grep "gray(255)" | awk '{print $2}'`)
num=${#bboxArr[*]}
echo $num

for ((i=0; i<num; i++)); do
bbox=${bboxArr[$i]}
echo $bbox
magick image.jpg -crop $bbox +repage result-$i.jpg
done

```

![result-0](https://github.com/user-attachments/assets/588c3d75-b08b-4ab6-b90d-72f1e9cab4be)

![result-1](https://github.com/user-attachments/assets/aa7ca61f-f850-4f16-b357-b706abc54c49)

![result-2](https://github.com/user-attachments/assets/06a184fc-22be-4e81-8d75-42502fb24e26)

**Sorry, I do not know Windows syntax for the FOR loop processing. A Windows user should be able to convert my code.**

**Reply by dmcnaugh15** [reply]
Wow.  That's some powerful stuff.  I do have an Ubuntu boot partition, so will try that tomorrow.  Will also wait for awhile to see if someone has a Windows script.  Thanks again for everything @fmw42.

**Comment by dmcnaugh15** [comment]
Okay, I'm attempting to convert this to a windows script step-by-step, using mogrify, since this is going to be a batch job.

**STEP 1 (Thresholding): Appears to be working**

- Script:
```magick mogrify -path C:\Users\dmcna\Desktop\Editing\Trimmed -format png -color-threshold "srgb(240,240,240)-srgb(255,255,255)" -negate -bordercolor black -border 1 *```

- Result:  Generates .png image in the output folder that looks like a black and white mask with the borders turned black (see attached photo - "image.png"). I assume this is what STEP 2 needs to work from.


**STEP 2 (Connected components / boundary cropping identification): Not working - CMD line errors**  

- Although I ultimately want this be be in a mogrify format, I moved the generated x.png file from the output folder back into the main folder, and started with your script.  
- Script (simply copied and pasted yours as follows):
  ```
  bboxArr=(`magick x.png \
  -type bilevel \
  -define connected-components:verbose=true \
  -define connected-components:keep-top=3 \
  -define connected-components:exclude-header=true \
  -connected-components 4 null: | grep "gray(255)" | awk '{print $2}'`)
  num=${#bboxArr[*]}
  echo $num
  ```

- Result: a lot of cmd line errors (see screenshot): 


**STEP 3 (For loop / original doc cropping): Will attempt after STEP 2**




PNG OUTPUT FILE (STEP 1):
![Image](https://github.com/user-attachments/assets/0e631f46-d4ed-473c-9774-4c397f69cc3d)


CMD LINE ERRORS (STEP 2):
![CMD line errors](https://github.com/user-attachments/assets/4432ca70-6f30-41f1-99c8-34bc734acd66)

**Reply by fmw42** [reply]
You are getting the correct image from step one, but the threshold values may change from image to image if the white border on your input is not in that range.

Windows FOR loops and variable setting does not work any way the same as for in Unix. You need a Windows expert to help you here.

You should test your Windows script using magick, not magick mogrify until that works. The FOR looping may need extensive changes as it likely does not work in magick mogrify.  You will likely have to write another outer FOR loop over each image to process them with magick (not magick mogrify).

**Reply by dmcnaugh15** [reply]
Sounds good.  I think you're right... even if I get it working with windows and mogrify, it still may not be safe to run as a batch with hundreds of images, as it may result in unwanted cropping of images.  I feel like this would be something for AI, but not sure we're there yet.

**Reply by fmw42** [reply]
Get it working for you in Unix and test on a variety of images, one at a time.  If that seems to work, then try to get it converted to Windows. (User snibgo, may be able to help, when he reads this post. He may have Windows script to do the same. See his web site at http://im.snibgo.com, though I did not find it on quick review.)

Are your borders always white? If so, check a few images to see what the range of white is? You can use some color picker tool to inspect the images.

**Comment by snibgo** [comment]
My page [Subimage rectangles](http://im.snibgo.com/subimrect.htm) may be useful. For example, we first add a white border to the image, then chop it up:
```
magick hillx3.jpg -bordercolor White -border 20 x.png

call %PICTBAT%rectSubimages x.png hill.png White 5 10c
```
This creates 3 output images, hill-0.png, hill-1.png and hill-2.png, which are the trimmed results.

**Comment by dmcnaugh15** [comment]
@snibgo Thank you for helping.   I checked out your Submanage Rectangles page; it is both overwhelming and awesome. Great work there. I'll be digesting it for awhile.

As far as this case goes, the first step worked normally, and created the x.png file.  The second step is generating an error, and I'm wondering if you can help me with that.  Attached are screenshots of the Windows File Explorer and CMD line.  The attempted scripts I've used are as follows:

1.
    ```
    call %PICTBAT%rectSubimages x.png Example.png White 5 10c
    ```

2.
    ```
    magick call %PICTBAT%rectSubimages x.png Example.png White 5 10c
    ```

![1  Explorer](https://github.com/user-attachments/assets/8cab17af-42e4-4b97-b7f7-3bf21271fd34)
![2  CMD Line](https://github.com/user-attachments/assets/58fda418-1b2b-408b-b424-a9eb3d67eb1c)

**Comment by snibgo** [comment]
I wrote:
```
call %PICTBAT%rectSubimages x.png hill.png White 5 10c
```
But you used:
```
magick call %PICTBAT%rectSubimages x.png Example.png White 5 10c
```
Why did you add `magick` to the front?

**Comment by dmcnaugh15** [comment]
Sorry, I should have explained.  I tried it according to script 1, which was what you provided.  That didn't work, so then I tried adding magick as a long-shot, which displays my noob status quite well.  But, in the screenshot of the cmd line, you can see I did use yours.

**Comment by snibgo** [comment]
What directory did you download `rectSubimages.bat` into? This should be either (a) your current directory, which is `C:\Users\dmcna\Desktop\Editing`, or (b) some directory that is on your system path or (c) some directory, and you set the environment variable `PICTBAT` to point to that directory.

If you have done one of these, then "`call %PICTBAT%rectSubimages`" should work.

**Comment by dmcnaugh15** [comment]
Oh my gosh.  I didn't even realize I was supposed to install it.  Will do when I go home from work tonight.  Sorry about that.  I'm sure it will work.

**Comment by snibgo** [comment, marked as answer]
The simplest method for you is to download http://im.snibgo.com/imsnibgoBats.zip which contains all my published BAT scripts and C code, and unzip that into some new directory such as `C:\snibgostuff` or whatever you like, then put that directory on your system path (eg System Properties, click on "Environment Variables", "System variables", "Path", click "Edit", and add `C:\snibgostuff`).

**Comment by dmcnaugh15** [comment]
Amazing!  It's like magic.  Thank u!!! 


Here is the result:

![Example-0](https://github.com/user-attachments/assets/6c9a6f26-3286-437b-8514-03733cfe7af4)
![Example-1](https://github.com/user-attachments/assets/b21d41ba-592c-457e-b544-f25cb2fb5735)
![Example-2](https://github.com/user-attachments/assets/59bb12b9-a6f2-414e-adfe-cc7caf464d6b)

**Comment by snibgo** [comment]
Good stuff. We can trim off those thin white triangles at the edges like this:
```
magick hill-0.png -bordercolor White -border 1 -define trim:percent-background=0% -fuzz 10% -trim out.png
```

**Comment by dmcnaugh15** [comment]
Nice.  Images 1 and 2 trim perfectly.  Image 3 trims too much off the top , probably because of all the white pixels in the sky.  I tried adjusting the fuzz percentage, from 1% all the way to 20%, but nothing really gets it right.  It's not a huge deal at this point.  I have a lot of scanned images that are probably going to need to be manually cropped anyway, since they have captions written below them.  Attached are the output files.  

I do have a followup question though.... what exactly does the percentage mean for things like fuzz or deskew?  For example, sometimes I have to use deskew 80% to get something to rotate by a couple degrees. So, the percent there isn't referring to something like degrees.  Similarly, with fuzz, the percent doesn't seem to refer to a percentage of space to crop.

![out](https://github.com/user-attachments/assets/14629e5d-3503-486e-b7bd-d14be8446ccb)
![out2](https://github.com/user-attachments/assets/49eb2950-b3b0-4c06-ae71-ae83c8970809)
![out3](https://github.com/user-attachments/assets/7c0a3bb2-888b-40d1-bc77-8cff85766c95)

**Reply by fmw42** [reply]
Lower the fuzz value.  It is happening because the sky near the top of the input is too white.

**Comment by dmcnaugh15** [comment]
I tried values ranging from 1-10% and none of them work.   The result is a vertical slice.  For example, attached is the result with a 5% fuzz value. 
![out3](https://github.com/user-attachments/assets/dc884256-90a5-4563-9dbd-83da709b2e58)

**Reply by fmw42** [reply]
Post the input if not the one above?  What are you trying to do exactly with the input?  Why does the input need trimming after separating them from the original containing the several images? Sorry, I know little about Windows scripting. But did you double the % to %% on Windows?

**Reply by dmcnaugh15** [reply]
I can confirm double % symbols were used.  This was the line:

```
magick Example-2.png -bordercolor White -border 1 -define trim:percent-background=0% -fuzz 5%% -trim out3.png
```

**Reply by snibgo** [reply]
> trim:percent-background=0% -fuzz 5%%

You need to double the other % sign.

**Reply by dmcnaugh15** [reply]
Sorry, I pasted an old line above that still just had only 1 (not 2) percent symbols.  This was actually the line I used.  And, I just used it again for confirmation:

```
magick Example-2.png -bordercolor White -border 1 -define trim:percent-background=0%% -fuzz 5%% -trim out3.png
```

**Comment by snibgo** [comment]
`-fuzz` determines whether one colour is "close enough" to another colour. Imagine a colour cube, with axes R, B and B. If two colours are within a certain difference, they are considered the same for the purposes of `-trim` and other operations. The diagonals of the colour cube (for example from black to white) are 100%, so `-fuzz 10%` is one-tenth of this distance.

The number in `-deskew 40` is a threshold for the radon transform, whatever that is. See https://en.wikipedia.org/wiki/Radon_transform. 

For the sky and land image, this works well for me:
```
magick skyLand.png -bordercolor White -border 1 -define trim:percent-background=0% -fuzz 10% -trim out.png
```
5% is also good, but shows a sliver of white at top of left edge.

If doing this in a BAT file, double every %.

**Comment by dmcnaugh15** [comment]
@snibgo Thanks for the explanation.  So, if -deskew is a threshold, then is it true that a higher percent value will be less likely to make a change?

Really odd, I'm cutting and pasting your script with the same results as before.  Can't think of a way to explain that.  Here's the line:

```
magick Example-3.png -bordercolor White -border 1 -define trim:percent-background=0%% -fuzz 10%% -trim out3.png
```
